### PR TITLE
Dem new inlet introducing rv

### PR DESCRIPTION
**Pull request by GuillermoCasas**
**Description**
This is another step towards refactoring and expanding the DEM inlet. Here I have added the possibility of using an arbitrary, piecewise linear distribution for the radii usigng the random_variable utility. The necessary data is passed through the json parameters. In fact, the whole project parameters are passed to the inlet, something that I think is justified for large, complex objects like this one (although we can discuss it). 

In principle, nothing changes for the previously existing distributions. Also, if no parameters are passed to the inlet, an empty one is created. This has been done to avoid API problems. It is not a pretty solution, but, please, bear in mind that it is only an intermediate step. Also, take into account the urgency of having this tool available for current projects.

Future steps include:
- Creating an inlet manager to manage the different inlet objects (there will be one per sub-model part)
- Creating an abstract 'properties' object to contain all the information required to create a DEM element (or an array of elements) deterministically, i.e., no statistical information, which should be dealt with outside.
- Using the random_variable utility for all cases (normal, lognormal...) as well
- Removing the inlet properties from the mdpa as much as possible, including all the necessary information in the parameters
- Creating an abstract StatisticalPropertiesSampler object to sample all the stochastic properties used to create new elements. For that, it will use a corresponding array of random_variable objects. This will allow to deal with highly complex situations such as the radii distribution being conditioned to the material type distribution etc. 

**Changelog**
- Added the piecewise linear radii distribution in the create_and_destroy
- Created a test that measures the empirical density function obtained through the inlet (iterating until it matches the desired distribution up to a tolerance)
- Added the possibility to pass the Parameters to the inlet

### CODE REVIEW

File: applications/DEMApplication/custom_utilities/inlet.cpp [hunk 1]
```diff
@@ -42,8 +42,12 @@ namespace Kratos {
 
     /// Constructor
 
-    DEM_Inlet::DEM_Inlet(ModelPart& inlet_modelpart, const int seed): mInletModelPart(inlet_modelpart)
-    {
+    DEM_Inlet::DEM_Inlet(ModelPart& inlet_modelpart, const int seed):
+                        DEM_Inlet(inlet_modelpart, Parameters(R"({"dem_inlets_settings" : {}})"), seed){}
```
Comment (maceligueta) [inline applications/DEMApplication/custom_utilities/inlet.cpp]:
```suggestion
                        DEM_Inlet(inlet_modelpart, Parameters(R"({})"), seed){}
```
I think "dem_inlets_settings" can stay outside

File: applications/DEMApplication/python_scripts/DEM_analysis_stage.py [hunk 2]
```diff
@@ -475,7 +475,7 @@ def SolverSolve(self):
     def SetInlet(self):
         if self.DEM_parameters["dem_inlet_option"].GetBool():
             #Constructing the inlet and initializing it (must be done AFTER the self.spheres_model_part Initialize)
-            self.DEM_inlet = DEM_Inlet(self.dem_inlet_model_part, self.seed)
+            self.DEM_inlet = DEM_Inlet(self.dem_inlet_model_part, self.DEM_parameters, self.seed)
```
Comment (maceligueta) [inline applications/DEMApplication/python_scripts/DEM_analysis_stage.py]:
```suggestion
            self.DEM_inlet = DEM_Inlet(self.dem_inlet_model_part, self.DEM_parameters["dem_inlets_settings"], self.seed)
```

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py [hunk 3]
```diff
@@ -550,7 +550,7 @@ def SetInlet(self):
             # Note that right now only inlets of a single type are possible.
             # This should be generalized.
             if self.project_parameters["type_of_inlet"].GetString() == 'VelocityImposed':
-                self.DEM_inlet = DEM_Inlet(self.dem_inlet_model_part, self.seed)
+                self.DEM_inlet = DEM_Inlet(self.dem_inlet_model_part, self.project_parameters['dem_parameters'], self.seed)
```
Comment (maceligueta) [inline applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py]:
```suggestion
                self.DEM_inlet = DEM_Inlet(self.dem_inlet_model_part, self.project_parameters['dem_parameters']["dem_inlets_settings"], self.seed)
```

**Comment by maceligueta** [comment]
Isn't it better to just pass the inlets settings? Not blocking, just proposing.

**Review by farrufat-cimne — COMMENTED** [review]
maybe its an indentation view glitch but I dont see where that  "i" is used in the loop

File: applications/DEMApplication/tests/random_variable_tests_files/random_variable.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		self.areas = np.zeros(len(self.heights) - 1)
+		self.trapezoid_indices = range(len(self.areas))
+		self.Normalize()
+		super().__init__(parameters)
+
+	def InterpolateHeight(self, x):
+		if x < self.boundaries[0] or x > self.boundaries[-1]:
+			raise Exception('x =' + str(x) + 'is out of bounds (' + str([self.boundaries[0], self.boundaries[-1]]) + ')')
+
+		for i, x_boundary in enumerate(self.boundaries):
```
Comment (farrufat-cimne) [inline applications/DEMApplication/tests/random_variable_tests_files/random_variable.py:33]: sorry, the comment was meant to be here

File: applications/DEMApplication/custom_utilities/inlet.cpp (hunk 1 above)
Comment (GuillermoCasas) [reply applications/DEMApplication/custom_utilities/inlet.cpp]: But it is required in the other constructor to have these settings...The initial idea (as explained in the description) was to pass the whole parameters object to the inlet. However, I agree to only pass the inlet settings for now. We can postpone this discussion until the inlets_manager exists.

File: applications/DEMApplication/tests/random_variable_tests_files/random_variable.py (hunk 4 above)
Comment (GuillermoCasas) [reply applications/DEMApplication/tests/random_variable_tests_files/random_variable.py:33]: Immediately after (e.g., `b1 = self.heights[i - 1]`). it is used outside the loop. Remember that in python the variables survive their scope.

**Comment by GuillermoCasas** [comment]
I had to change the foillowing line:

`materials_file_abs_path = os.path.join(self.main_path, str(adapted_to_current_os_relative_path))`

by forcing the second argumnet to be a string; otherwise my version of python (3.5) gave an error:

> TypeError: join() argument must be str or bytes, not 'PosixPath'

File: applications/DEMApplication/custom_utilities/inlet.cpp [hunk 5]
```diff
@@ -42,8 +42,12 @@ namespace Kratos {
 
     /// Constructor
 
-    DEM_Inlet::DEM_Inlet(ModelPart& inlet_modelpart, const int seed): mInletModelPart(inlet_modelpart)
-    {
+    DEM_Inlet::DEM_Inlet(ModelPart& inlet_modelpart, const int seed):
+                        DEM_Inlet(inlet_modelpart, Parameters(R"({}})"), seed){}
```
Comment (maceligueta) [inline applications/DEMApplication/custom_utilities/inlet.cpp]:
```suggestion
                        DEM_Inlet(inlet_modelpart, Parameters(R"({})"), seed){}
```

**Review by maceligueta — APPROVED** [review]

**Pull request merged**